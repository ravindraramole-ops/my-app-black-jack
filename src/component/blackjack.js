import React from 'react';
import '../index.css';

//Renders card object
function Card(props){
    if(props.suit === "hearts"){
        return(
            <div className="hands">
                {props.notation} 
                <span>&hearts;</span>
            </div>
        );
    }
    if(props.suit === "spades"){
        return(
            <div className="hands">
                {props.notation} 
                <span>&spades;</span>
            </div>
        );
    }
    if(props.suit === "clubs"){
        return(
            <div className="hands">
                {props.notation} 
                <span>&clubs;</span>
            </div>
        );
    }
    if(props.suit === "diamonds"){
        return(
            <div className="hands">
                {props.notation} 
                <span>&diams;</span>
            </div>
        );
    }
    if(props.suit === "face-down"){
        return(
            <div className="hands">
                ####
            </div>
        );
    }
}

//Renders dealer and player hands
class Hand extends React.Component{
    renderHand(){
        const rows = [];
        for(let i=0; i<this.props.hand.length; i++) {
            //initial dace down card for dealer
            if(this.props.faceDownCard && i===1){
                rows.push(<Card key={this.props.hand[i].notation+this.props.hand[i].suit} notation={this.props.hand[i].notation} suit={'face-down'} />);
            }else{
                rows.push(<Card key={this.props.hand[i].notation+this.props.hand[i].suit} notation={this.props.hand[i].notation} suit={this.props.hand[i].suit} />);
            }
        }
        return rows;
    }

    render(){
        return(
            <div className="flex-center flex-row">{this.renderHand()}</div>
        );
    }
}

//React component game
class Game extends React.Component {
    constructor(props){
        super(props);
        this.cards = generateDeck(1);
        let playerHand = [], dealerHand = [];
        //Initializing the initial hands for player and dealer
        playerHand.push(getRandomCard(this.cards))
        playerHand.push(getRandomCard(this.cards))
        dealerHand.push(getRandomCard(this.cards))
        dealerHand.push(getRandomCard(this.cards))

        let playerHandVal = getHandValue(playerHand);
        let dealerHandVal = getHandValue(dealerHand);

        //if player hits 21 in the initial hand then BlackJack
        if(playerHandVal === 21){
            this.state = {
                playerHand: playerHand,
                dealerHand: dealerHand,
                gameStatus: {status:"BlackJack", playerMode: false},
                playerHandValue: playerHandVal,
                dealerHandValue: dealerHandVal
            }
        }else{
            this.state = {
                playerHand: playerHand,
                dealerHand: dealerHand,
                gameStatus: {status:"In Progress", playerMode: true},
                playerHandValue: playerHandVal,
                dealerHandValue: dealerHandVal
            }
        }
    }

    //functionality to handle game option hit
    onHit(){
        //if player already busted or standed then no action
        if(!this.state.gameStatus.playerMode){
            return;
        }
        let playerHandVal = getHandValue(this.state.playerHand);
        if(playerHandVal<=21){
            let playerHand = this.state.playerHand.slice(); //Maintaining immutibility of the state object by getting a copy of object and then modifying
            let newCard = getRandomCard(this.cards);
            playerHand.push(newCard);   //added new card to the player's hand
            playerHandVal = getHandValue(playerHand);
            //update the state according to the player's hand value
            if(playerHandVal > 21){
                this.setState({playerHand: playerHand, gameStatus: {status: "Busted", playerMode:false}, playerHandValue: playerHandVal});
            } else{
                this.setState({playerHand: playerHand, playerHandValue: playerHandVal});
            }
        }
    }

    //functionality to handle game option stand
    onStand(){
        if(!this.state.gameStatus.playerMode){
            return;
        }
        //dealer auto play logic
        let dealerHandVal = getHandValue(this.state.dealerHand);
        //dealer stands on 17, no card added to dealer's hand after the hand value reaches or crosses 17
        while(dealerHandVal<17){
            let dealerHand = this.state.dealerHand.slice();
            let randomCard = getRandomCard(this.cards);
            dealerHand.push(randomCard);
            dealerHandVal = getHandValue(dealerHand);
            this.setState({dealerHand: dealerHand, dealerHandValue: dealerHandVal});
        }
        //compatring the player and dealer hand to get the game status
        let gameStatus = compareHands({playerHand: this.state.playerHand, dealerHand: this.state.dealerHand});
        this.setState({gameStatus: {status: gameStatus, playerMode:false}});
    }

    //Resetting the complete game status to initial state
    onReset(){
        this.cards = generateDeck(1);
        let playerHand = [], dealerHand = [];
        playerHand.push(getRandomCard(this.cards))
        playerHand.push(getRandomCard(this.cards))
        dealerHand.push(getRandomCard(this.cards))
        dealerHand.push(getRandomCard(this.cards))

        let playerHandVal = getHandValue(playerHand);
        let dealerHandVal = getHandValue(dealerHand);

        if(playerHandVal === 21){
            this.setState({
                playerHand: playerHand,
                dealerHand: dealerHand,
                gameStatus: {status:"BlackJack", playerMode: false},
                playerHandValue: playerHandVal,
                dealerHandValue: dealerHandVal
            });
        }else{
            this.setState({
                playerHand: playerHand,
                dealerHand: dealerHand,
                gameStatus: {status:"In Progress", playerMode: true},
                playerHandValue: playerHandVal,
                dealerHandValue: dealerHandVal
            });
        }
    }

    render(){
        let status = this.state.gameStatus.status;
        let faceDownCard = this.state.gameStatus.playerMode ? true : false;

        return(
            <div className="container">
                <div className="flex-center flex-column">
                    <div className="flex-center flex-row">
                        <h1>Dealer</h1>
                    </div>
                    <Hand hand={this.state.dealerHand} faceDownCard={faceDownCard} />                    
                </div>

                <div className="flex-center flex-column">
                    <div className="flex-center flex-row">
                        <div className="game-options">
                            <button id="btnstand" className="btn" onClick={()=>this.onStand(this)}>Stand</button>
                            <button id="btnhit" className="btn" onClick={()=>this.onHit(this)}>Hit</button>
                            <button id="btnreset" className="btn" onClick={()=>this.onReset(this)}>Reset</button>                                                       
                        </div>
                    </div>
                    <div className="status flex-center flex-row">                        
                        <h1>Game Status: {status}</h1>                                                                                
                    </div>
                </div>                                

                <div className="flex-center flex-column">
                    <div className="flex-center flex-row">
                        <h1>Player</h1>
                    </div>
                    <Hand hand={this.state.playerHand} faceDownCard={false}/>                    
                </div>
            </div>
        );
    }
}

//generates and returns a card deck
function generateDeck(numOfDecks = 1){
    let cards = [];
    for(let i=0; i<numOfDecks; i++){
        for(let j=0; j<13; j++){
            let notation = '', val;
            if(j<9){
                notation += j+2;
                val = j+2;
            }
            if(j===9){
                notation = 'J';
                val = 10
            }
            if(j===10){
                notation = 'Q';
                val = 10
            }
            if(j===11){
                notation = 'K';
                val = 10
            }
            if(j===12){
                notation = 'A';
                val = 11
            }
            cards.push({
                notation: notation, suit: 'hearts', value: val,
            })
            cards.push({
                notation: notation, suit: 'spades', value: val,
            })
            cards.push({
                notation: notation, suit: 'clubs', value: val,
            })
            cards.push({
                notation: notation, suit: 'diamonds', value: val,
            })
        }
    }
    return cards;
}

//draws random card from the card deck and updates the card deck
function getRandomCard(cards){
    let len = cards.length;
    let randomIdx = Math.floor((Math.random()*1000)%len);
    let retObj = cards[randomIdx];
    cards.splice(randomIdx, 1);
    return retObj;
}

//calculates hand value and considers ace value 1 or 11 as per the hand benefits
function getHandValue(handObj){
    let handVal = 0, numOfAces = 0;
    for(let i=0; i<handObj.length; i++){
        if(handObj[i].notation === 'A'){
            numOfAces += 1;
        }
        handVal += handObj[i].value;
        if(handVal > 21 && numOfAces > 0){
            handVal -= numOfAces * 10;
            numOfAces -= 1
        }
    }
    return handVal;
}

//compares player and dealer hand to decide the winner
function compareHands(props){
    let playerHandVal = getHandValue(props.playerHand);
    let dealerHandVal = getHandValue(props.dealerHand);
    if(playerHandVal > 21){
        return "Busted";
    }
    if(playerHandVal <=21 && dealerHandVal <=21 && playerHandVal === dealerHandVal){
        return "Tie (Push)";
    }
    if( (playerHandVal <=21 && playerHandVal > dealerHandVal) || dealerHandVal > 21){
        return "Won";
    }
    if( (dealerHandVal <=21 && dealerHandVal > playerHandVal) || playerHandVal > 21){
        return "Lost";
    }
}

export {Game, Hand, Card, generateDeck, getRandomCard, getHandValue, compareHands};