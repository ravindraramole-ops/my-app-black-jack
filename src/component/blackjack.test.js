import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {Game, Hand, Card, generateDeck, getRandomCard, getHandValue, compareHands} from './blackjack';

configure({ adapter: new Adapter() });

describe("Card deck", () => {
    let cardDeckNum = 1;
    let cards = generateDeck(cardDeckNum);
    it('Length', () => {
      expect(cards).toHaveLength(cardDeckNum * 52);
    });
    it('First card', () => {
      let initailCard = {notation: '2', suit: 'hearts', value: 2};
      expect(cards[0]).toEqual(initailCard);
    });
    it('Last card', () => {
      let finalCard = {notation: 'A', suit: 'diamonds', value: 11};
      expect(cards[cards.length-1]).toEqual(finalCard);
    });
    it('Draw random card', () => {
      let cardsNum = cards.length;
      let card = getRandomCard(cards);
      let newCardsNum = cards.length;
      expect(newCardsNum).toBe(cardsNum-1);
    })
});

describe("Get hand value", () => {

  it('Number Cards', () =>{
    let handObj = [];
    handObj.push({notation: "2", suit: "hearts", value: 2});
    handObj.push({notation: "4", suit: "clubs", value: 4});
    handObj.push({notation: "8", suit: "spades", value: 8});
    expect(getHandValue(handObj)).toBe(14);
  })

  it('Face Cards', () =>{
    let handObj = [];
    handObj.push({notation: "K", suit: "diamonds", value: 10});
    handObj.push({notation: "J", suit: "hearts", value: 10});
    handObj.push({notation: "8", suit: "spades", value: 8});
    expect(getHandValue(handObj)).toBe(28);
  })

  it('Ace card value 11', () =>{
    let handObj = [];
    handObj.push({notation: "A", suit: "clubs", value: 11});
    handObj.push({notation: "K", suit: "hearts", value: 10});
    expect(getHandValue(handObj)).toBe(21);
  })

  it('Ace card value 1', () =>{
    let handObj = [];
    handObj.push({notation: "A", suit: "clubs", value: 11});
    handObj.push({notation: "8", suit: "spades", value: 8});
    handObj.push({notation: "6", suit: "hearts", value: 6});
    expect(getHandValue(handObj)).toBe(15);
  })

})

describe("Compare hands", () => {

  it("Player busted", () => {
    let playerHand = [];
    playerHand.push({notation: "K", suit: "spades", value: 10});
    playerHand.push({notation: "4", suit: "spades", value: 4});
    playerHand.push({notation: "8", suit: "hearts", value: 8});
    let dealerHand = [];
    dealerHand.push({notation: "9", suit: "spades", value: 9});
    dealerHand.push({notation: "4", suit: "spades", value: 4});
    expect(compareHands({playerHand: playerHand, dealerHand: dealerHand})).toBe("Busted");
  })

  it("Dealer busted", () => {
    let dealerHand = [];
    dealerHand.push({notation: "K", suit: "spades", value: 10});
    dealerHand.push({notation: "4", suit: "spades", value: 4});
    dealerHand.push({notation: "8", suit: "hearts", value: 8});
    let playerHand = [];
    playerHand.push({notation: "9", suit: "spades", value: 9});
    playerHand.push({notation: "4", suit: "spades", value: 4});
    expect(compareHands({playerHand: playerHand, dealerHand: dealerHand})).toBe("Won");
  })

  it("Player high hand", () => {
    let dealerHand = [];
    dealerHand.push({notation: "4", suit: "spades", value: 4});
    dealerHand.push({notation: "8", suit: "hearts", value: 8});
    let playerHand = [];
    playerHand.push({notation: "9", suit: "spades", value: 9});
    playerHand.push({notation: "4", suit: "spades", value: 4});
    playerHand.push({notation: "4", suit: "spades", value: 4});
    expect(compareHands({playerHand: playerHand, dealerHand: dealerHand})).toBe("Won");
  })

  it("Dealer high hand", () => {
    let dealerHand = [];
    dealerHand.push({notation: "4", suit: "spades", value: 4});
    dealerHand.push({notation: "8", suit: "hearts", value: 8});
    dealerHand.push({notation: "9", suit: "spades", value: 9});
    let playerHand = [];
    playerHand.push({notation: "9", suit: "spades", value: 9});
    playerHand.push({notation: "4", suit: "spades", value: 4});
    expect(compareHands({playerHand: playerHand, dealerHand: dealerHand})).toBe("Lost");
  })

  it("Tie", () => {
    let dealerHand = [];
    dealerHand.push({notation: "4", suit: "spades", value: 4});
    dealerHand.push({notation: "8", suit: "hearts", value: 8});
    let playerHand = [];
    playerHand.push({notation: "9", suit: "spades", value: 9});
    playerHand.push({notation: "3", suit: "spades", value: 3});
    expect(compareHands({playerHand: playerHand, dealerHand: dealerHand})).toBe("Tie (Push)");
  })

})

describe("BlackJack Game", () => {
  
  it('Game container renders correctly', () => {
    const gamewrapper = shallow(<Game />);
    expect(gamewrapper).toMatchSnapshot();
  });

  it('Hand container renders correctly', () => {
    let handObj = [];
    handObj.push({notation: "A", suit: "clubs", value: 11});
    handObj.push({notation: "8", suit: "spades", value: 8});
    handObj.push({notation: "6", suit: "hearts", value: 6});
    const gamewrapper = shallow(<Hand hand={handObj} faceDownCard={false}/>);
    expect(gamewrapper).toMatchSnapshot();
  });

  it('Initial hands', () => {
    const gamewrapper = shallow(<Game />);
    expect(gamewrapper.state('playerHand')).toHaveLength(2);
    expect(gamewrapper.state('dealerHand')).toHaveLength(2);
  });

  it('Player hits', () => {
    const gamewrapper = shallow(<Game />);
    const buttonElement = gamewrapper.find('#btnhit');
    let initialCards = 2;
    while(gamewrapper.state('gameStatus').playerMode){
      buttonElement.simulate('click');
      initialCards += 1;
      expect(gamewrapper.state('playerHand')).toHaveLength(initialCards);
    }
  });

  it('Dealer stands on 17', () => {
    const gamewrapper = shallow(<Game />);
    const buttonElement = gamewrapper.find('#btnstand');
    let initialCards = 2;
    let initialDealerHandValue = gamewrapper.state('dealerHandValue');
    if(initialDealerHandValue >= 17){
      buttonElement.simulate('click');
      expect(gamewrapper.state('dealerHand')).toHaveLength(initialCards);
    } else {
      buttonElement.simulate('click');
      expect(gamewrapper.state('dealerHandValue')).toBeLessThan(27);
    }
  });

  it('Declare Winner', () => {
    const gamewrapper = shallow(<Game />);
    const buttonElement = gamewrapper.find('#btnhit');
    const buttonElement2 = gamewrapper.find('#btnstand');
    buttonElement.simulate('click');
    buttonElement.simulate('click');
    buttonElement2.simulate('click');
    let playerHandVal = gamewrapper.state('playerHandValue');
    let dealerHandVal = gamewrapper.state('dealerHandValue');
    let status = gamewrapper.state('gameStatus').status;
    if( playerHandVal == 21 && playerHandVal.length == 2){
      expect(status).toBe("BlackJack");
    }else if( playerHandVal > 21){
      expect(status).toBe("Busted");
    }else if(playerHandVal > dealerHandVal){
      expect(status).toBe("Won");
    }else if(dealerHandVal > 21){
      expect(status).toBe("Won");
    }else if(dealerHandVal > playerHandVal){
      expect(status).toBe("Lost");
    }else if(dealerHandVal == playerHandVal){
      expect(status).toBe("Tie (Push)");
    }
  });

});