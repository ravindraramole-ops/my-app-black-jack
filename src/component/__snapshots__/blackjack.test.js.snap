// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlackJack Game Game container renders correctly 1`] = `ShallowWrapper {}`;

exports[`BlackJack Game Hand container renders correctly 1`] = `ShallowWrapper {}`;

exports[`BlackJack Game container renders correctly 1`] = `ShallowWrapper {}`;

exports[`BlackJack Renders correctly 1`] = `ShallowWrapper {}`;
